src/main/java/com/gtnewhorizon/structurelib/structure/StructureUtility.java

```diff
@@ -444,24 +517,82 @@ public static PlaceResult survivalPlaceBlock(ItemStack stack, NBTMode nbtMode, N
-            return PlaceResult.REJECT;
-        if (!s.takeOne(stack, false))
-            // this is bad! probably an exploit somehow. Let's nullify the block we just placed instead
+        if (mode != REMOVING) {
+            if (mode == BUILDING) {
+                if (!StructureLibAPI.isBlockTriviallyReplaceable(world, x, y, z, actor)) {
+                    return PlaceResult.SKIP;
+                }
+            }
+            if (mode == UPDATING) {
+                if (!world.isAirBlock(x, y, z)) {
+                    if (world.getTileEntity(x, y, z) instanceof TileEntity) {
+                        return PlaceResult.SKIP;
+                    }
+                    Block blockInWorld = world.getBlock(x, y, z);
+                    ItemStack blockInWorldItem = new ItemStack(blockInWorld);
+                    blockInWorldItem.setItemDamage(blockInWorld.getDamageValue(world, x, y, z));
+                    if (blockInWorldItem.isItemEqual(stack)) {
+                        // This is the block we are looking for, move along
+                        return PlaceResult.SKIP;
+                    }
+                }
+            }
+            if (!assumeStackPresent && !s.takeOne(stack, true)) {
+                if (chatter != null) chatter.accept(
+                        new ChatComponentTranslation(
+                                "structurelib.autoplace.error.no_item_stack",
+                                stack.func_151000_E()));
+                return PlaceResult.REJECT;
+            }
+            if (mode == UPDATING && !world.isAirBlock(x, y, z) && checkPermission(world, actor, x, y, z)) {
+                // We're in updating mode and there's a block in our way (grr), so let's remove whatever is there
+                // ...and drop it to the player
+                ItemStack toGive = new ItemStack(world.getBlock(x, y, z));
+                toGive.setItemDamage(world.getBlock(x, y, z).getDamageValue(world, x, y, z));
+                if (!world.setBlock(x, y, z, Blocks.air, 0, 4) || !tryGiveOrDropItem(world, actor, toGive)) {
+                    // We either weren't able to give the player the block that was there, or we weren't able to remove
+                    // the block, either way, we need to reject the update. Also, since we hid the block update from
+                    // the client in an effort to make the exchange as seamless as possible, we need to now notify
+                    // the client if we did successfully remove the block.
+                    if (world.isAirBlock(x, y, z)) world.notifyBlockChange(x, y, z, Blocks.air);
+                    return PlaceResult.REJECT;
+                }
+            }
+            if (!stack.copy()
+                    .tryPlaceItemIntoWorld(actor, world, x, y, z, ForgeDirection.UP.ordinal(), 0.5f, 0.5f, 0.5f)) {
+                return PlaceResult.REJECT;
+            }
+        }
+        if (mode == REMOVING || !s.takeOne(stack, false)) {
```

Comment:
Please split this if block into two. One for the hypothetical take from item source failed case and the other for commencing removing.